src/parsers/manifestjson.js

```diff
@@ -37,28 +38,58 @@ export default class ManifestJSONParser {
     this.isValid = this._validate();
   }
 
+  errorLookup(error) {
+    // This is the default message.
+    var baseObject = messages.MANIFEST_JSON_INVALID;
+    var overrides = {
+      message: `${error.dataPath} ${error.message}`,
+    };
+
+    if (error.keyword === 'required') {
+      baseObject = messages.MANIFEST_FIELD_REQUIRED;
+    } else if (error.keyword === 'type') {
+      baseObject = messages.MANIFEST_FIELD_INVALID;
+    }
+
+    // Arrays can be extremely verbose, this tries to make them a little
+    // more sane. Using a regex because there will likely be more as we
+    // expand the schema.
+    var match = error.dataPath.match(/^\/(permissions)\/([\d+])/);
+    if (match) {
+      baseObject = messages[`MANIFEST_${match[1].toUpperCase()}`];
+      overrides.message = singleLineString`/${match[1]}: Unknown ${match[1]}
+          "${error.data}" at ${match[2]}.`;
+    }
+
+    return Object.assign({}, baseObject, overrides);
+  }
+
   _validate() {
+    // Not all messages returned by the schema are fatal to Firefox, messages
+    // that are just warnings should be added to this array.
+    var warnings = [messages.MANIFEST_PERMISSIONS.code];
+
     var isValid = validate(this.parsedJSON);
     if (!isValid) {
-      log.debug('Schema Validation errors', validate.errors);
+      log.debug('Schema Validation messages', validate.errors);
+      var errorsFound = [];
+
       for (let error of validate.errors) {
-        var description;
-        var errorData = {
-          code: 'MANIFEST_JSON_INVALID',
-          message: `${error.dataPath}: ${error.message}`,
-          file: 'manifest.json',
-        };
-
-        // If a required prop is missing, introspect the schema for its
-        // description.
-        if (error.keyword === 'required') {
-          description = error.schema[error.params.missingProperty].description;
-        } else {
-          description = error.parentSchema.description;
+        // Ensure that we only add one error on a field. This runs the risk
+        // of hiding errors, but means that we can aim to get to a more
+        // helpful error in the case of some rather verbose schema errors.
+        if (errorsFound.indexOf(error.dataPath) > -1) {
+          continue;
         }
 
-        errorData.description = description || 'MISSING_SCHEMA_DESCRIPTION';
-        this.collector.addError(errorData);
+        var message = this.errorLookup(error);
+        errorsFound.push(error.dataPath);
+
+        if (warnings.indexOf(message.code) > -1) {
```

Comment:
Can you use [`includes`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes) here instead? It's a nicer API.